pkg/shardddl/optimism/keeper.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optimism
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+)
+
+// LockKeeper used to keep and handle DDL lock conveniently.
+// The lock information do not need to be persistent, and can be re-constructed from the shard DDL info.
+type LockKeeper struct {
+	mu    sync.RWMutex
+	locks map[string]*Lock // lockID -> Lock
+}
+
+// NewLockKeeper creates a new LockKeeper instance.
+func NewLockKeeper() *LockKeeper {
+	return &LockKeeper{
+		locks: make(map[string]*Lock),
+	}
+}
+
+// TrySync tries to sync the lock.
+func (lk *LockKeeper) TrySync(info Info, sts []SourceTables) (string, []string, error) {
+	var (
+		lockID = genDDLLockID(info)
+		l      *Lock
+		ok     bool
+	)
+
+	lk.mu.Lock()
+	defer lk.mu.Unlock()
+
+	if l, ok = lk.locks[lockID]; !ok {
+		lk.locks[lockID] = NewLock(lockID, info.Task, info.TableInfoBefore, sts)
```

Comment:
YES, it will _try_ to add `sts` twice, but I think it's acceptable, is it right?

The one for `NewLock` is in order to assign the initial schema for tables, another one for `TrySync` is in order to support the newly added tables (like `CREATE TABLE`).